File: GridView/VirtualRadGridView/VirtualRadGridViewCS/VirtualRadGridViewSortingFiltering.Core/VirtualRadGridView.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Reflection;
using System.Text;
using System.Windows.Forms;
using Telerik.WinControls.Data;
using Telerik.WinControls.UI;

namespace VirtualRadGridViewSortingFiltering.Core
{
    public class VirtualRadGridView : RadGridView
    {
        private RadWaitingBar loadingOverlay;

        public ItemSource ItemsSource
        {
            get { return (this.GridViewElement as VirtualRadGridViewElement).ItemsSource; }
        }

        public object VirtualDataSource
        {
            get { return this.ItemsSource.DataSource; }
            set
            {
                this.FilterDescriptors.Clear();
                this.SortDescriptors.Clear();
                this.GroupDescriptors.Clear();
                this.ItemsSource.DataSource = value;
                this.Initialize();
            }
        }

        public bool ShowLoadingOverlay { get; set; }

        public bool AutomaticallyRetreiveCellValues { get; set; }

        public bool AutomaticallyPushCellValues { get; set; }

        public RadWaitingBar LoadingOverlay
        {
            get { return this.loadingOverlay; }
        }

        public VirtualRadGridView()
            : base()
        {
            this.EnablePaging = true;
            this.EnableSorting = true;
            this.EnableFiltering = true;
            this.EnableGrouping = true;
            this.VirtualMode = true;
            this.AutomaticallyPushCellValues = true;
            this.AutomaticallyRetreiveCellValues = true;
            this.MasterTemplate.PagingBeforeGrouping = true;
            this.ShowLoadingOverlay = true;

            this.loadingOverlay = this.CreateLoadingOverlay();
            this.loadingOverlay.Parent = this;
            this.loadingOverlay.Dock = DockStyle.None;
            this.loadingOverlay.WaitingStyle = Telerik.WinControls.Enumerations.WaitingBarStyles.Dash;
            this.loadingOverlay.Visible = false;
        }

        protected virtual RadWaitingBar CreateLoadingOverlay()
        {
            return new RadWaitingBar();
        }

        protected override void OnBindingContextChanged(EventArgs e)
        {
            base.OnBindingContextChanged(e);
            this.Initialize();
        }

        protected virtual void Initialize()
        {
            if (this.BindingContext == null)
            {
                return;
            }

            if (this.RowCount == this.ItemsSource.Count && this.ColumnCount == this.ItemsSource.BoundProperties.Count)
            {
                return;
            }

            this.InitializeRowsAndColumns();
        }

        protected virtual void InitializeRowsAndColumns()
        {
            if (!this.IsLoaded || this.VirtualDataSource == null)
            {
                return;
            }

            this.BeginUpdate();
            this.InitializeColumns();
            this.InitializeRows();
            this.EndUpdate();
        }

        protected virtual void InitializeColumns()
        {
            if (this.ColumnCount != this.ItemsSource.BoundProperties.Count)
            {
                this.ColumnCount = this.ItemsSource.BoundProperties.Count;
                for (int i = 0; i < this.Columns.Count; i++)
                {
                    PropertyDescriptor prop = this.ItemsSource.BoundProperties[i];
                    GridViewDataColumn newColumn = GridViewHelper.AutoGenerateGridColumn(prop.PropertyType, null);
                    newColumn.HeaderText = prop.DisplayName;
                    newColumn.Name = prop.Name;

                    this.Columns.RemoveAt(i);
                    this.Columns.Insert(i, newColumn);
                }
            }
        }

        protected virtual void InitializeRows()
        {
            if (this.RowCount != this.ItemsSource.Count)
            {
                this.RowCount = this.ItemsSource.Count;
            }
        }

        protected override void WireEvents()
        {
            base.WireEvents();

            this.SortChanged += VirtualRadGridView_SortChanged;
            this.FilterChanged += VirtualRadGridView_FilterChanged;
            this.CellEndEdit += VirtualRadGridView_CellEndEdit;
            this.ItemsSource.OperationCompleted += ItemsSource_OperationCompleted;
            this.ItemsSource.OperationStarted += ItemsSource_OperationStarted;
        }

        protected override void UnWireEvents()
        {
            base.UnWireEvents();

            this.SortChanged -= VirtualRadGridView_SortChanged;
            this.FilterChanged -= VirtualRadGridView_FilterChanged;
            this.CellEndEdit -= VirtualRadGridView_CellEndEdit;
            this.ItemsSource.OperationCompleted -= ItemsSource_OperationCompleted;
            this.ItemsSource.OperationStarted -= ItemsSource_OperationStarted;
        }

        protected override void Dispose(bool disposing)
        {
            base.Dispose(disposing);

            if (disposing)
            {
                this.ItemsSource.Dispose();
            }
        }

        protected override RadGridViewElement CreateGridViewElement()
        {
            return new VirtualRadGridViewElement();
        }

        protected override void OnResize(EventArgs e)
        {
            base.OnResize(e);
            if (this.loadingOverlay.Visible)
            {
                this.PositionLoadingOverlay();
            }
        }

        protected virtual void PositionLoadingOverlay()
        {
            int x = (this.Width / 2) - (this.loadingOverlay.Width / 2);
            int y = (this.Height / 2) - (this.loadingOverlay.Height / 2);
            this.loadingOverlay.Location = new System.Drawing.Point(x, y);
        }

        protected virtual void StartWaiting()
        {
            if (!this.ShowLoadingOverlay)
            {
                return;
            }

            this.PositionLoadingOverlay();

            this.loadingOverlay.Visible = true;
            this.loadingOverlay.StartWaiting();
        }

        protected virtual void StopWaiting()
        {
            this.loadingOverlay.StopWaiting();
            this.loadingOverlay.Visible = false;
        }

        void ItemsSource_OperationStarted(object sender, ItemSourceOperationEventArgs e)
        {
            this.StartWaiting();
        }

        void ItemsSource_OperationCompleted(object sender, ItemSourceOperationCompletedEventArgs e)
        {
            if (e.Canceled)
            {
                this.StopWaiting();
                return;
            }

            if ((e.OperationType & ItemSourceOperation.Filtering) == ItemSourceOperation.Filtering)
            {
                this.InitializeRowsAndColumns();
            }

            if ((e.OperationType & ItemSourceOperation.Sorting) == ItemSourceOperation.Sorting)
            {
                this.MasterTemplate.DataView.Refresh();
            }

            this.StopWaiting();
        }

        void VirtualRadGridView_FilterChanged(object sender, GridViewCollectionChangedEventArgs e)
        {
            if (!this.IsInEditMode)
            {
                this.PerformFilter();
            }
        }

        void VirtualRadGridView_CellEndEdit(object sender, GridViewCellEventArgs e)
        {
            if (e.Row is GridViewFilteringRowInfo)
            {
                this.PerformFilter();
            }
        }

        void VirtualRadGridView_SortChanged(object sender, GridViewCollectionChangedEventArgs e)
        {
            this.PerformSort();
        }

        protected virtual void PerformFilter(bool force = false)
        {
            this.ItemsSource.PerformOperation(ItemSourceOperation.Filtering, this.FilterDescriptors, force);
        }

        protected virtual void PerformSort(bool force = false)
        {
            this.ItemsSource.PerformOperation(ItemSourceOperation.Sorting, this.SortDescriptors, force);
        }

        protected override void OnCellValueNeeded(object sender, GridViewCellValueEventArgs e)
        {
            if (!this.AutomaticallyRetreiveCellValues)
            {
                base.OnCellValueNeeded(sender, e);
                return;
            }

            GridViewCellValueEventArgsEx args = new GridViewCellValueEventArgsEx(e, e.ColumnIndex, e.RowIndex);
            if (this.GroupDescriptors.Count == 0)
            {
                int rowIndex = this.GetRowIndexByPage(e.RowIndex);
                args.Value = this.ItemsSource.GetValue(this.ItemsSource[rowIndex], e.ColumnIndex);
            }
            else
            {
                args.Value = this.ItemsSource.GetValue(args.RowInfo.DataBoundItem, e.ColumnIndex);
            }

            base.OnCellValueNeeded(sender, args);
            e.Value = args.Value;
        }

        protected override void OnCellValuePushed(object sender, GridViewCellValueEventArgs e)
        {
            if (!this.AutomaticallyPushCellValues)
            {
                base.OnCellValuePushed(sender, e);
                return;
            }

            GridViewCellValueEventArgsEx args = new GridViewCellValueEventArgsEx(e, e.ColumnIndex, e.RowIndex);
            if (this.GroupDescriptors.Count == 0)
            {
                int rowIndex = this.GetRowIndexByPage(e.RowIndex);
                this.ItemsSource.SetValue(this.ItemsSource[rowIndex], e.ColumnIndex, e.Value);
            }
            else
            {
                this.ItemsSource.SetValue(args.RowInfo.DataBoundItem, e.ColumnIndex, e.Value);
            }

            base.OnCellValuePushed(sender, args);
        }

        public virtual int GetRowIndexByPage(int rowIndex)
        {
            int newRowIndex = rowIndex;
            if (this.MasterTemplate.PageIndex >= 1)
            {
                newRowIndex = newRowIndex + (this.MasterTemplate.PageSize * this.MasterTemplate.PageIndex);
            }

            return newRowIndex;
        }
    }
}

File: GridView/VirtualRadGridView/VirtualRadGridViewSortingFilteringAfterQ3_2015/VirtualRadGridViewSortingFilteringAfterQ3_2015VB/VirtualRadGridViewSortingFiltering.Core/VirtualRadGridView.vb

Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Reflection
Imports System.Text
Imports System.Windows.Forms
Imports Telerik.WinControls.UI

Namespace VirtualRadGridViewSortingFiltering.Core
    Public Class VirtualRadGridView
        Inherits RadGridView
        Private m_loadingOverlay As RadWaitingBar

        Public ReadOnly Property ItemsSource() As ItemSource
            Get
                Return TryCast(Me.GridViewElement, VirtualRadGridViewElement).ItemsSource
            End Get
        End Property

        Public Property VirtualDataSource() As Object
            Get
                Return Me.ItemsSource.DataSource
            End Get
            Set(value As Object)
                Me.FilterDescriptors.Clear()
                Me.SortDescriptors.Clear()
                Me.GroupDescriptors.Clear()
                Me.ItemsSource.DataSource = value
                Me.Initialize()
            End Set
        End Property

        Public Property ShowLoadingOverlay() As Boolean
            Get
                Return m_ShowLoadingOverlay
            End Get
            Set(value As Boolean)
                m_ShowLoadingOverlay = Value
            End Set
        End Property
        Private m_ShowLoadingOverlay As Boolean

        Public Property AutomaticallyRetreiveCellValues() As Boolean
            Get
                Return m_AutomaticallyRetreiveCellValues
            End Get
            Set(value As Boolean)
                m_AutomaticallyRetreiveCellValues = Value
            End Set
        End Property
        Private m_AutomaticallyRetreiveCellValues As Boolean

        Public Property AutomaticallyPushCellValues() As Boolean
            Get
                Return m_AutomaticallyPushCellValues
            End Get
            Set(value As Boolean)
                m_AutomaticallyPushCellValues = Value
            End Set
        End Property
        Private m_AutomaticallyPushCellValues As Boolean

        Public ReadOnly Property LoadingOverlay() As RadWaitingBar
            Get
                Return Me.m_loadingOverlay
            End Get
        End Property

        Public Sub New()
            MyBase.New()
            Me.EnablePaging = True
            Me.EnableSorting = True
            Me.EnableFiltering = True
            Me.EnableGrouping = True
            Me.VirtualMode = True
            Me.AutomaticallyPushCellValues = True
            Me.AutomaticallyRetreiveCellValues = True
            Me.MasterTemplate.PagingBeforeGrouping = True
            Me.ShowLoadingOverlay = True

            Me.m_loadingOverlay = Me.CreateLoadingOverlay()
            Me.m_loadingOverlay.Parent = Me
            Me.m_loadingOverlay.Dock = DockStyle.None
            Me.m_loadingOverlay.WaitingStyle = Telerik.WinControls.Enumerations.WaitingBarStyles.Dash
            Me.m_loadingOverlay.Visible = False
        End Sub

        Protected Overridable Function CreateLoadingOverlay() As RadWaitingBar
            Return New RadWaitingBar()
        End Function

        Protected Overrides Sub OnBindingContextChanged(e As EventArgs)
            MyBase.OnBindingContextChanged(e)
            Me.Initialize()
        End Sub

        Protected Overridable Sub Initialize()
            If Me.BindingContext Is Nothing Then
                Return
            End If

            If Me.RowCount = Me.ItemsSource.Count AndAlso Me.ColumnCount = Me.ItemsSource.BoundProperties.Count Then
                Return
            End If

            Me.InitializeRowsAndColumns()
        End Sub

        Protected Overridable Sub InitializeRowsAndColumns()
            If Not Me.IsLoaded OrElse Me.VirtualDataSource Is Nothing Then
                Return
            End If

            Me.BeginUpdate()
            Me.InitializeColumns()
            Me.InitializeRows()
            Me.EndUpdate()
        End Sub

        Protected Overridable Sub InitializeColumns()
            If Me.ColumnCount <> Me.ItemsSource.BoundProperties.Count Then
                Me.ColumnCount = Me.ItemsSource.BoundProperties.Count
                For i As Integer = 0 To Me.Columns.Count - 1
                    Dim prop As PropertyDescriptor = Me.ItemsSource.BoundProperties(i)
                    Dim newColumn As GridViewDataColumn = GridViewHelper.AutoGenerateGridColumn(prop.PropertyType, Nothing)
                    newColumn.HeaderText = prop.DisplayName
                    newColumn.Name = prop.Name

                    Me.Columns.RemoveAt(i)
                    Me.Columns.Insert(i, newColumn)
                Next
            End If
        End Sub

        Protected Overridable Sub InitializeRows()
            If Me.RowCount <> Me.ItemsSource.Count Then
                Me.RowCount = Me.ItemsSource.Count
            End If
        End Sub

        Protected Overrides Sub WireEvents()
            MyBase.WireEvents()

            AddHandler Me.SortChanged, AddressOf VirtualRadGridView_SortChanged
            AddHandler Me.FilterChanged, AddressOf VirtualRadGridView_FilterChanged
            AddHandler Me.CellEndEdit, AddressOf VirtualRadGridView_CellEndEdit
            AddHandler Me.ItemsSource.OperationCompleted, AddressOf ItemsSource_OperationCompleted
            AddHandler Me.ItemsSource.OperationStarted, AddressOf ItemsSource_OperationStarted
        End Sub

        Protected Overrides Sub UnWireEvents()
            MyBase.UnWireEvents()

            RemoveHandler Me.SortChanged, AddressOf VirtualRadGridView_SortChanged
            RemoveHandler Me.FilterChanged, AddressOf VirtualRadGridView_FilterChanged
            RemoveHandler Me.CellEndEdit, AddressOf VirtualRadGridView_CellEndEdit
            RemoveHandler Me.ItemsSource.OperationCompleted, AddressOf ItemsSource_OperationCompleted
            RemoveHandler Me.ItemsSource.OperationStarted, AddressOf ItemsSource_OperationStarted
        End Sub

        Protected Overrides Sub Dispose(disposing As Boolean)
            MyBase.Dispose(disposing)

            If disposing Then
                Me.ItemsSource.Dispose()
            End If
        End Sub

        Protected Overrides Function CreateGridViewElement() As RadGridViewElement
            Return New VirtualRadGridViewElement()
        End Function

        Protected Overrides Sub OnResize(e As EventArgs)
            MyBase.OnResize(e)
            If Me.m_loadingOverlay.Visible Then
                Me.PositionLoadingOverlay()
            End If
        End Sub

        Protected Overridable Sub PositionLoadingOverlay()
            Dim x As Integer = (Me.Width / 2) - (Me.m_loadingOverlay.Width / 2)
            Dim y As Integer = (Me.Height / 2) - (Me.m_loadingOverlay.Height / 2)
            Me.m_loadingOverlay.Location = New System.Drawing.Point(x, y)
        End Sub

        Protected Overridable Sub StartWaiting()
            If Not Me.ShowLoadingOverlay Then
                Return
            End If

            Me.PositionLoadingOverlay()

            Me.m_loadingOverlay.Visible = True
            Me.m_loadingOverlay.StartWaiting()
        End Sub

        Protected Overridable Sub StopWaiting()
            Me.m_loadingOverlay.StopWaiting()
            Me.m_loadingOverlay.Visible = False
        End Sub

        Private Sub ItemsSource_OperationStarted(sender As Object, e As ItemSourceOperationEventArgs)
            Me.StartWaiting()
        End Sub

        Private Sub ItemsSource_OperationCompleted(sender As Object, e As ItemSourceOperationCompletedEventArgs)
            If e.Canceled Then
                Me.StopWaiting()
                Return
            End If

            If (e.OperationType And ItemSourceOperation.Filtering) = ItemSourceOperation.Filtering Then
                Me.InitializeRowsAndColumns()
            End If

            If (e.OperationType And ItemSourceOperation.Sorting) = ItemSourceOperation.Sorting Then
                Me.MasterTemplate.DataView.Refresh()
            End If

            Me.StopWaiting()
        End Sub

        Private Sub VirtualRadGridView_FilterChanged(sender As Object, e As GridViewCollectionChangedEventArgs)
            If Not Me.IsInEditMode Then
                Me.PerformFilter()
            End If
        End Sub

        Private Sub VirtualRadGridView_CellEndEdit(sender As Object, e As GridViewCellEventArgs)
            If TypeOf e.Row Is GridViewFilteringRowInfo Then
                Me.PerformFilter()
            End If

        End Sub

        Private Sub VirtualRadGridView_SortChanged(sender As Object, e As GridViewCollectionChangedEventArgs)
            Me.PerformSort()
        End Sub

        Protected Overridable Sub PerformFilter(Optional force As Boolean = False)
            Me.ItemsSource.PerformOperation(ItemSourceOperation.Filtering, Me.FilterDescriptors, force)
        End Sub

        Protected Overridable Sub PerformSort(Optional force As Boolean = False)
            Me.ItemsSource.PerformOperation(ItemSourceOperation.Sorting, Me.SortDescriptors, force)
        End Sub

        Protected Overrides Sub OnCellValueNeeded(sender As Object, e As GridViewCellValueEventArgs)
            If Not Me.AutomaticallyRetreiveCellValues Then
                MyBase.OnCellValueNeeded(sender, e)
                Return
            End If

            Dim args As New GridViewCellValueEventArgsEx(e, e.ColumnIndex, e.RowIndex)
            If Me.GroupDescriptors.Count = 0 Then
                Dim rowIndex As Integer = Me.GetRowIndexByPage(e.RowIndex)
                args.Value = Me.ItemsSource.GetValue(Me.ItemsSource(rowIndex), e.ColumnIndex)
            Else
                args.Value = Me.ItemsSource.GetValue(args.RowInfo.DataBoundItem, e.ColumnIndex)
            End If

            MyBase.OnCellValueNeeded(sender, args)
            e.Value = args.Value
        End Sub

        Protected Overrides Sub OnCellValuePushed(sender As Object, e As GridViewCellValueEventArgs)
            If Not Me.AutomaticallyPushCellValues Then
                MyBase.OnCellValuePushed(sender, e)
                Return
            End If

            Dim args As New GridViewCellValueEventArgsEx(e, e.ColumnIndex, e.RowIndex)
            If Me.GroupDescriptors.Count = 0 Then
                Dim rowIndex As Integer = Me.GetRowIndexByPage(e.RowIndex)
                Me.ItemsSource.SetValue(Me.ItemsSource(rowIndex), e.ColumnIndex, e.Value)
            Else
                Me.ItemsSource.SetValue(args.RowInfo.DataBoundItem, e.ColumnIndex, e.Value)
            End If

            MyBase.OnCellValuePushed(sender, args)
        End Sub

        Public Overridable Function GetRowIndexByPage(rowIndex As Integer) As Integer
            Dim newRowIndex As Integer = rowIndex
            If Me.MasterTemplate.PageIndex >= 1 Then
                newRowIndex = newRowIndex + (Me.MasterTemplate.PageSize * Me.MasterTemplate.PageIndex)
            End If

            Return newRowIndex
        End Function
    End Class
End Namespace

'=======================================================
'Service provided by Telerik (www.telerik.com)
'Conversion powered by NRefactory.
'Twitter: @telerik
'Facebook: facebook.com/telerik
'=======================================================
